Handle bindings in substructure of patterns with type ascriptions

pnkfelix opened the pull request:
This attempts to follow the outline described by @nikomatsakis [here](https://github.com/rust-lang/rust/issues/47184#issuecomment-420041056). Its a bit more complicated than expected for two reasons: 

 1. In general it handles sets of type ascriptions, because such ascriptions can be nested within patterns
 2.  It has a separate types in the HAIR, `PatternTypeProjections` and `PatternTypeProjection`, which are analogues to the corresponding types in the MIR.

The main reason I added the new HAIR types was because I am worried that the current implementation is inefficent, and asymptotically so: It makes copies of vectors as it descends the patterns, even when those accumulated vectors are never used.

Longer term, I would like to used a linked tree structure for the `PatternTypeProjections` and `PatternTypeProjection`, and save the construction of standalone vectors for the MIR types. I didn't want to block landing this on that hypoethetical revision; but I figured I could at least make the future change easier by differentiating between the two types now.

Oh, one more thing: This doesn't attempt to handle `ref x` (in terms of ensuring that any necessary types are ascribed to `x` in that scenario as well). We should open an issue to investigate supporting that as well. But I didn't want to block this PR on that future work.

Fix #54570

pnkfelix committed:
Remove `PlaceContext` from API of `mir::Visitor::visit_projection_elem`.

It is unused, and would not make sense to maintain in the commits later in this PR.

pnkfelix committed:
restrict scope of methods that are only called within mod tree of parent.

pnkfelix committed:
Refactoring: added `PatternTypeAnnotation` wrapper around `UserTypeAnnotation` to ease future changes.

pnkfelix committed:
Added `mir::UserTypeProjection`, a stub for a structure that projects *into* a given UserTypeAnnotation.

(That is, it will pull out some component type held or referenced by
the type annotation.)

Note: this still needs to actually do projection itself. That comes in
a later commit

pnkfelix committed:
Checkpoint: Added abstraction over collection of projections into user type.

I did not think I would need this in the MIR, but in general local
decls are going to need to support this. (That, or we need to be able
define a least-upper-bound for a collection of types encountered via
the pattern compilation.)

pnkfelix committed:
Add the actual chain of projections to `UserTypeProjection`.

Update the existing NLL `patterns.rs` test accordingly.

includes changes addressing review feedback:

 * Added example to docs for `UserTypeProjections` illustrating how we
   build up multiple projections when descending into a pattern with
   type ascriptions.

 * Adapted niko's suggested docs for `UserTypeProjection`.

 * Factored out `projection_ty` from more general `projection_ty_core`
   (as a drive-by, made its callback an `FnMut`, as I discovered later
   that I need that).

 * Add note to docs that `PlaceTy.field_ty(..)` does not normalize its result.

 * Normalize as we project out `field_ty`.

rust-highfive commented:
r? @zackmdavis

(rust_highfive has picked a reviewer for you, use r? to override)

pnkfelix commented:
r? @nikomatsakis

pnkfelix commented:
Oh, also: the `patterns.rs` should be extended with cases that use `enum` that are analogous to the current ones that use `struct`.

If I don't get a chance to add that to this PR before its reviewed, then I will file a follow-up `E-needs-test` issue documenting the desired test.

pnkfelix commented:
Oh, and I just realized that I should have looked more carefully at @nikomatsakis 's note a couple more times; namely, he anticipated the issue with `PlaceProjection` carrying the `Ty` for things like fields; that's why he pointed us at [sanitize_projection](https://github.com/rust-lang/rust/blob/5aee959e9f7525e95c1f7231cc6dacc0f66a7f6a/src/librustc_mir/borrow_check/nll/type_check/mod.rs#L446)

My own code has its own hacked up version of that which only covers a subset of the cases, here:

https://github.com/rust-lang/rust/blob/e8c1c8b0c4b0fa89388771dba41532a0d3e1dfa5/src/librustc/mir/tcx.rs#L47-L69

comparing the above to:

https://github.com/rust-lang/rust/blob/a66dc8a1489d2818b5e63f18e1464be79da1f137/src/librustc_mir/borrow_check/nll/type_check/mod.rs#L641-L708

my main takeaway is that ... my hacked up version does not handle closures nor generators. But ... does it need to??? Can those arise in this context (of user type ascriptions)?

pnkfelix committed:
Update mir-opt tests to reflect change to `AscribeUserType` to carry a `UserTypeProjection`.

rust-highfive commented:
The job `x86_64-gnu-llvm-5.0` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/444845325) ([raw log](https://api.travis-ci.org/v3/job/444845325/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain

```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

pnkfelix commented:
Hmm i didn’t see that in my (stage1) builds. Will investigate.

pnkfelix commented:
Okay I've derived a standalone test case from the `rustc` source. Investigating further.

pnkfelix commented:
Hmm I must be missing a normalization step somewhere in there; my minimized test seems to pivot on whether one is using a `type` alias or not.

pnkfelix commented:
and it also seems to pivot on the particular order I use for the fields in my local definition of `struct Spanned<T>` ...? I'm starting to think that my hacked up code for projecting out the fields is more buggy than I realized.

pnkfelix commented:
@nikomatsakis I think its definitely ready for partial review.

I've already tagged the bit of code that I *suspect* is most suspicious / in-need-of-revision. Namely, the way I've dealt with `field_ty` lookup.

But a review of the overall architecture would still be very useful at this point.

pnkfelix commented:
(but in case its not clear: It also is buggy. I need to fix it so that the [above test](https://github.com/rust-lang/rust/pull/55274#issuecomment-432271840) works.)

pnkfelix commented:
Oh! I bet the bug is this: 
https://github.com/rust-lang/rust/blob/d3c91ae31a0676b2575b83f47252f892b6c25f24/src/librustc_mir/build/matches/mod.rs#L579-L583

namely, the index `j` is just the index of the subpattern in the HAIR; there's *no* connection between that and the index of the corresponding field. So dumb!

pnkfelix commented:
(and I think the fix is pretty simple; I really should have reviewed HAIR more carefully before I wrote this code.)

pnkfelix committed:
regression test for ICE I encountered in my patch.

pnkfelix commented:
(implemented aforementioned fix, and added the reduced test. Unfortunately I couldn't figure out a way to generalize the test to actually appear relevant to the change here without causing *other* checks to fire and thus mask the ICE.)

nikomatsakis commented on src/librustc/mir/mod.rs:
```diff
@@ -2438,6 +2438,42 @@ EnumTypeFoldableImpl! {
+/// A collection of projections into user types.
+///
+/// They are projections because a binding can occur a part of a
+/// parent pattern that has been ascribed a type.
```
It would be nice to give some Rust code examples here. I guess something like

```rust
let ((x, _): (T1, _)): (T2, _) = ...
```

is what you have in mind?

pnkfelix replied:
I cannot decide whether something like `let a @ Thing1(_, b @ Thing2(_, _): Inner): Outer` would be clearer or not.

pnkfelix replied:
okay added example to docs.

nikomatsakis commented on src/librustc/mir/mod.rs:
```diff
@@ -2474,14 +2474,34 @@ impl<'tcx> UserTypeProjections<'tcx> {
 
-#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash, RustcEncodable, RustcDecodable)]
+#[derive(Clone, Debug, PartialEq, Eq, Hash, RustcEncodable, RustcDecodable)]
 pub struct UserTypeProjection<'tcx> {
```
I think we should definitely add a comment to this type explaining what-the-heck is going on. Here is my attempt:

---

Encodes a user-supplied type annotation -- the type annotation is determined by applying the given list of projections to an underlying base type. Often, the projection-element list is empty, in which case this simply encodes a type. But in the case of complex patterns we may need to apply only a "subset" of the type to a variable, in which case the `projs` vector is used.

Example:

- `let x: T = ...` -- here, the `projs` vector is empty
- `let (x, _): T = ...` -- here, the `projs` vector would contain the `0` field, indicating that the type of `x` is determined by finding the type of the `0` field from `T`

pnkfelix replied:
Oy you are correct, I had a "todo" for myself in a commit message to add docs and then I think I either ignored it, or deleted it during a rebase

pnkfelix replied:
added adaptation of above text to the docs for `UserTypeProjection`.

nikomatsakis commented on src/librustc/mir/tcx.rs:
```diff
@@ -44,9 +44,41 @@ impl<'a, 'gcx, 'tcx> PlaceTy<'tcx> {
+            (PlaceTy::Downcast { adt_def, substs, variant_index }, _) => {
+                let variant_def = &adt_def.variants[variant_index];
+                let field_def = &variant_def.fields[f.index()];
+                field_def.ty(tcx, substs)
```
Maybe worth noting that the resulting type here is not normalized?

pnkfelix replied:
add note to doc for `field_ty` that the resulting ty is not normalized.

nikomatsakis commented on src/librustc/mir/tcx.rs:
```diff
@@ -44,9 +44,41 @@ impl<'a, 'gcx, 'tcx> PlaceTy<'tcx> {
+    /// `handle_field` callback is used to map a `Field` to its `Ty`;
+    /// in most cases one can just use the `Ty` stored inline, which
+    /// is passed as the third parameter to the `handle_field`
+    /// callback.
```
Maybe we want a convenient wrapper and an "inner" fn that takes the closure?

nikomatsakis commented on src/librustc_mir/borrow_check/nll/type_check/mod.rs:
```diff
@@ -1011,6 +1017,14 @@ impl<'a, 'gcx, 'tcx> TypeChecker<'a, 'gcx, 'tcx> {
+                let mut projected_ty = PlaceTy::from_ty(ty);
+                for proj in &user_ty.projs {
+                    projected_ty = projected_ty.projection_ty(
+                        tcx, proj, |this, field, &()| this.field_ty(tcx, field));
```
We will need to normalize here, I think

pnkfelix replied:
Okay I think this will serve as a basis for a test showing that we need to normalize as we project the fields:

```rust
struct Single<T> { value: T }

trait Indirect {
    type U;
}

impl<'a> Indirect for &'a u32 {
    type U = &'a u32;
}

fn static_to_a_to_static_through_struct_2<'a>(_x: &'a u32) -> &'static u32 {
    let Single { value: y }: Single<<&'a u32 as Indirect>::U> = Single { value: &22 };
    y //~ ERROR
}
```
```

pnkfelix replied:
(well, either that, or the example exposes bugs elsewhere. Ha ha ... ha. ha.)

pnkfelix replied:
right, the previous example looks an awful like an embedding of #54940. So maybe thats not going to get me anywhere.

pnkfelix replied:
(If it comes down to it, I'll happily put in the code to normalize in the callback itself. But I would like to spend a bit longer trying to identify a test case that the current PR breaks on...)

pnkfelix replied:
ah this one looks promising:

```rust
trait Indirect { type Assoc; }
struct StaticU32;
impl Indirect for StaticU32 { type Assoc = &'static u32; }

struct Single2<T: Indirect> { value: <T as Indirect>::Assoc }

fn static_no_initializer_2() {
    let x = 22;
    let Single2 { value: mut _y }: Single2<StaticU32>;
    _y = &x; //~ ERROR
}
```

nikomatsakis reviewed (commented):
First review — looks good but for the normalization question

pnkfelix committed:
Add test for normalization during field-lookup on patterns with ascribed types.

As a drive-by, also added test analogous to existing
static_to_a_to_static_through_tuple, but now apply to a struct instead
of a tuple.

rust-highfive commented:
The job `x86_64-gnu-llvm-5.0` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/446188831) ([raw log](https://api.travis-ci.org/v3/job/446188831/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
travis_time:end:00433623:start=1540476004754957098,finish=1540476061798990958,duration=57044033860
$ git checkout -qf FETCH_HEAD
travis_fold:end:git.checkout

Encrypted environment variables have been removed for security reasons.
See https://docs.travis-ci.com/user/pull-requests/#Pull-Requests-and-Security-Restrictions
$ export SCCACHE_BUCKET=rust-lang-ci-sccache2
$ export SCCACHE_REGION=us-west-1
Setting environment variables from .travis.yml
$ export SCCACHE_BUCKET=rust-lang-ci-sccache2
---
[00:07:14]    Compiling syntax_ext v0.0.0 (/checkout/src/libsyntax_ext)
[00:07:17] error[E0425]: cannot find value `el` in this scope
[00:07:17]   --> librustc/mir/tcx.rs:83:38
[00:07:17]    |
[00:07:17] 83 |         self.projection_ty_core(tcx, el, |_, _, ty| ty)
[00:07:17]    |                                      ^^ not found in this scope
[00:07:43] error: aborting due to previous error
[00:07:43] 
[00:07:43] For more information about this error, try `rustc --explain E0425`.
[00:07:44] error: Could not compile `rustc`.
[00:07:44] error: Could not compile `rustc`.
[00:07:44] 
[00:07:44] To learn more, run the command again with --verbose.
[00:07:44] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "build" "--target" "x86_64-unknown-linux-gnu" "-j" "4" "--release" "--locked" "--color" "always" "--features" " jemalloc" "--manifest-path" "/checkout/src/rustc/Cargo.toml" "--message-format" "json"
[00:07:44] expected success, got: exit code: 101
[00:07:44] thread 'main' panicked at 'cargo must succeed', bootstrap/compile.rs:1115:9
[00:07:44] travis_fold:end:stage0-rustc

[00:07:44] travis_time:end:stage0-rustc:start=1540476372763269974,finish=1540476535946552647,duration=163183282673


[00:07:44] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap build
[00:07:44] Build completed unsuccessfully in 0:03:39
[00:07:44] make: *** [all] Error 1
[00:07:44] Makefile:28: recipe for target 'all' failed

The command "stamp sh -x -c "$RUN_SCRIPT"" exited with 2.
travis_time:start:01ac38e0
$ date && (curl -fs --head https://google.com | grep ^Date: | sed 's/Date: //g' || true)
---
151412 ./src/tools/clang
149628 ./obj/build/bootstrap/debug/incremental
149120 ./src/llvm-emscripten/test
134172 ./obj/build/bootstrap/debug/incremental/bootstrap-32pr67l4sa8g0
134168 ./obj/build/bootstrap/debug/incremental/bootstrap-32pr67l4sa8g0/s-f61x03xniz-fn1ppn-1wbz4rzqyzyro
107668 ./obj/build/x86_64-unknown-linux-gnu/stage0/lib/rustlib/x86_64-unknown-linux-gnu/codegen-backends
104700 ./src/tools/lldb
93748 ./src/tools/clang/test
87540 ./obj/build/x86_64-unknown-linux-gnu/stage0-rustc
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

bors commented:
:umbrella: The latest upstream changes (presumably #55323) made this pull request unmergeable. Please resolve the merge conflicts.

pnkfelix committed:
Add intern table for `List<ProjectionElem<'tcx, (), ()>>`.

Also added alias `ProjectionKind<'tcx>` for `ProjectionElem<'tcx, (), ()>`.

pnkfelix committed:
Further foundational stuff on `ProjectionKind` before I add it to `AscribeUserType`.

pnkfelix committed:
add user_ty.projs support to `AscribeUserType`.

bors commented:
:umbrella: The latest upstream changes (presumably #53821) made this pull request unmergeable. Please resolve the merge conflicts.

nikomatsakis commented on src/librustc/mir/visit.rs:
```diff
@@ -797,8 +797,9 @@ macro_rules! make_mir_visitor {
             ) {
                 let UserTypeProjection {
                     ref $($mutability)* base,
+                    projs: _, // Note: Does not visit projection elems!
```
Is there any particular reason for this? I guess just because nobody has any reason to.

pnkfelix replied:
Well, the existing MIR visit methods take a `PlaceElem`, not a `ProjectionElem<(), ()>`. So it would have been a bit of extra complexity in the visit API to try to support it ... and to what end?  There isn't any recursive structure to visit, and you would probably want to know all about the surrounding context when inspecting any particular `ProejctionElem<(), ()>` within the `UserTypeProjection` ... so better to just do it explicitly in one's `visit_user_type_projection` method?

nikomatsakis commented:
r=me but needs rebase :'(

pnkfelix commented:
@bors p=1

(this is, like #55385, an RC2 "nice to have")

bors commented:
:lock: Merge conflict

This pull request and the master branch diverged in a way that cannot be automatically merged. Please rebase on top of the latest master branch, and let the reviewer approve again.

<details><summary>How do I rebase?</summary>

Assuming `self` is your fork and `upstream` is this repository, you can resolve the conflict following these steps:

1. `git checkout issue-54570-proj-path-into-pats-with-type-take-2` *(switch to your branch)*
2. `git fetch upstream master` *(retrieve the latest master)*
3. `git rebase upstream/master -p` *(rebase on top of it)*
4. Follow the on-screen instruction to resolve conflicts (check `git status` if you got lost).
5. `git push self issue-54570-proj-path-into-pats-with-type-take-2 --force-with-lease` *(update this PR)*

You may also read [*Git Rebasing to Resolve Conflicts* by Drew Blessing](http://blessing.io/git/git-rebase/open-source/2015/08/23/git-rebasing-to-resolve-conflicts.html) for a short tutorial.

Please avoid the ["**Resolve conflicts**" button](https://help.github.com/articles/resolving-a-merge-conflict-on-github/) on GitHub. It uses `git merge` instead of `git rebase` which makes the PR commit history more difficult to read.

Sometimes step 4 will complete without asking for resolution. This is usually due to difference between how `Cargo.lock` conflict is handled during merge and rebase. This is normal, and you should still perform step 5 to update this PR.

</details>

<details><summary>Error message</summary>

```text
Auto-merging src/librustc_mir/shim.rs
Auto-merging src/librustc_mir/build/mod.rs
Auto-merging src/librustc_mir/build/matches/mod.rs
CONFLICT (content): Merge conflict in src/librustc_mir/build/matches/mod.rs
Auto-merging src/librustc_mir/build/expr/as_place.rs
CONFLICT (content): Merge conflict in src/librustc_mir/build/expr/as_place.rs
Auto-merging src/librustc_mir/borrow_check/nll/type_check/mod.rs
CONFLICT (content): Merge conflict in src/librustc_mir/borrow_check/nll/type_check/mod.rs
Auto-merging src/librustc/ty/context.rs
Auto-merging src/librustc/mir/mod.rs
CONFLICT (content): Merge conflict in src/librustc/mir/mod.rs
Automatic merge failed; fix conflicts and then commit the result.

```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #55382) made this pull request unmergeable. Please resolve the merge conflicts.

pnkfelix commented:
@bors r=nikomatsakis p=1

bors commented:
:pushpin: Commit 639a3ffadcfec5165d434cbbf320678f69b01a5b has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis 639a3ffadcfec5165d434cbbf320678f69b01a5b -->

bors commented:
:hourglass: Testing commit 639a3ffadcfec5165d434cbbf320678f69b01a5b with merge 10f42cbde015c44a019e8b6dceca472a1532f36a...

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/builds/19838026), [status-travis](https://travis-ci.org/rust-lang/rust/builds/446951691?utm_source=github_status&utm_medium=notification)
Approved by: nikomatsakis
Pushing 10f42cbde015c44a019e8b6dceca472a1532f36a to master...

eggyal commented on src/librustc/mir/mod.rs:
```diff
@@ -2449,6 +2453,117 @@ EnumLiftImpl! {
+        use mir::ProjectionElem::*;
+
+        let base = self.base.fold_with(folder);
+        let projs: Vec<_> = self.projs
+            .iter()
+            .map(|elem| {
+                match elem {
+                    Deref => Deref,
+                    Field(f, ()) => Field(f.clone(), ()),
+                    Index(()) => Index(()),
+                    elem => elem.clone(),
+                }})
+            .collect();
+
+        UserTypeProjection { base, projs }
+    }
+
+    fn super_visit_with<Vs: TypeVisitor<'tcx>>(&self, visitor: &mut Vs) -> bool {
+        self.base.visit_with(visitor)
+        // Note: there's nothing in `self.proj` to visit.
```
I appreciate this was a very long time ago, but this seems odd to me.  If there was something in `self.projs` to fold, how can there be nothing to visit?  The lack of visiting `self.projs` remains today.
